Index: xen-4.8.1/tools/Rules.mk
===================================================================
--- xen-4.8.1.orig/tools/Rules.mk	2017-04-10 09:21:48.000000000 -0400
+++ xen-4.8.1/tools/Rules.mk	2017-08-16 17:21:00.427585349 -0400
@@ -151,7 +151,7 @@
 CFLAGS_libblktapctl = -I$(XEN_BLKTAP2)/control -I$(XEN_BLKTAP2)/include $(CFLAGS_xeninclude)
 SHDEPS_libblktapctl =
 LDLIBS_libblktapctl = $(SHDEPS_libblktapctl) $(XEN_BLKTAP2)/control/libblktapctl$(libextension)
-SHLIB_libblktapctl  = $(SHDEPS_libblktapctl) -Wl,-rpath-link=$(XEN_BLKTAP2)/control
+SHLIB_libblktapctl  = $(SHDEPS_libblktapctl) -lblktapctl
 else
 CFLAGS_libblktapctl =
 SHDEPS_libblktapctl =
